presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java

```diff
@@ -133,115 +200,440 @@ public boolean isFinished()
+        return state == State.NEEDS_INPUT && !partialOutput.isPresent();
     }
 
     @Override
     public void addInput(Page page)
     {
         checkState(needsInput(), "Operator is already finishing");
+        checkState(!partialOutput.isPresent());
         requireNonNull(page, "page is null");
 
-        long memorySize = 0;
-        for (Aggregator aggregate : aggregates) {
-            aggregate.processPage(page);
-            memorySize += aggregate.getEstimatedSize();
-        }
-        if (partial) {
-            systemMemoryContext.setBytes(memorySize);
-        }
-        else {
-            userMemoryContext.setBytes(memorySize);
-        }
+        partialOutput = processor.addInput(page);
     }
 
     @Override
     public Page getOutput()
     {
-        if (state != State.HAS_OUTPUT) {
+        if (partialOutput.isPresent()) {
+            Page output = partialOutput.get();
+            partialOutput = Optional.empty();
+            return output;
+        }
+
+        if (state != State.FINISHING) {
             return null;
         }
 
-        // project results into output blocks
-        List<Type> types = aggregates.stream().map(Aggregator::getType).collect(toImmutableList());
+        state = State.FINISHED;
+        return processor.getFinalOutput();
+    }
+
+    private interface Processor
+    {
+        Optional<Page> addInput(Page page);
+
+        Page getFinalOutput();
+    }
+
+    private static class SingleProcessor
+            implements Processor
+    {
+        private final LocalMemoryContext userMemoryContext;
+        private final List<InternalSingleAccumulator> accumulators;
+
+        public SingleProcessor(LocalMemoryContext userMemoryContext, List<InternalAccumulatorFactory> accumulatorFactories)
+        {
+            this.userMemoryContext = userMemoryContext;
+            this.accumulators = accumulatorFactories.stream()
+                    .map(InternalAccumulatorFactory::createSingleAccumulator)
+                    .collect(toImmutableList());
+        }
+
+        @Override
+        public Optional<Page> addInput(Page inputPage)
+        {
+            long memorySize = 0;
+            for (InternalSingleAccumulator accumulator : accumulators) {
+                accumulator.addInput(inputPage);
+                memorySize += accumulator.getEstimatedSize();
+            }
+            userMemoryContext.setBytes(memorySize);
+            return Optional.empty();
+        }
+
+        @Override
+        public Page getFinalOutput()
+        {
+            List<Block> finalResults = new ArrayList<>();
+            for (InternalSingleAccumulator accumulator : accumulators) {
+                BlockBuilder blockBuilder = accumulator.getFinalType().createBlockBuilder(null, 1);
+                accumulator.evaluateFinal(blockBuilder);
+                finalResults.add(blockBuilder.build());
+            }
+            return new Page(1, finalResults.stream().toArray(Block[]::new));
```

Comment:
`SelectedPositions#selectPositions` creates an array of `Block` from the beginning. This code and other places in this commit create a `List<Block>` first, fill it up, then convert it to an array. Why are these differences?